src/mlpack/methods/neat/gene.hpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * @file gene.hpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * Definition of the Gene classes.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_NEAT_GENE_HPP
+#define MLPACK_METHODS_NEAT_GENE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack{
+namespace neat /** NeuroEvolution of Augmenting Topologies */ {
+
+/**
+ * Connection gene class.
+ */
+class ConnectionGene
```

Comment:
Do you think you could draw the overall class connection infrastructure (what class depends on which, connection between classes) on a paper? It doesn't have to be fancy.

Reply:
![IMG_20190601_193935015_HDR](https://user-images.githubusercontent.com/40389657/58749566-3c70b380-84a5-11e9-8e23-98b67eaf6be0.jpg)
Do you mean something like this?

Reply:
Yes, thanks that is really helpful. So I think you will start with the the implementation of the Genome? You already have the interface for the ConnectionGene. which will be part of the Genome that also holds a list of Genes? If you like I can sketch up a high level interface?

Reply:
Yup, I'm working on the implementation of the Genome class right now. In the genome I am planning to maintain two lists of Connection Genes, where one will be an adjacency list to help with evaluating the phenomes and the other is a std::vector storing the genes sorted by innovation ID. As for the interface, besides what we discussed Input() and Output()) what else would the user need? The population would need to use a Mutate() and Crossover() function. I don't think there is much else.

Reply:
Maybe here instead of storing genes as a vector, you can just store pointers to genes sorted by innvoation ID.

Reply:
What would be the benefit to this over storing the genes directly? I think if we use pointers, I would have to dynamically allocate the genes, which would really complicate things with having to concern ourselves with memory management. Is there something I'm missing?

Reply:
I intend to find a way to reduce the space usage here. But, you are right about the issue of memory management, I completely missed out this point. And if I recall correctly, we usually don't do pointers here in mlpack.

Reply:
You could use ``ptr_vector``, or I guess you can also use move semantics to avoid another copy, but at the end you have to store the object somewhere.

> As for the interface, besides what we discussed Input() and Output()) what else would the user need? 


I was thinking about a more in depth internal class interface, this class will hold this object, this class is called by another class using this interface, etc.

Reply:
Oh, I understand. I can't think of much, besides making a class wrapping all the gene storage containers, which makes it easier to pass to the Phenome class? Maybe also a class that stores the "config" of the genome? Like the means for the weights and biases, input and output node counts, maximum complexity for pruning, etc. Once again we wouldn't have to use a lot of parameters to pass basic info to other methods and classes.

Reply:
Agreed, we should defiantly think about how we want to load a config, but don't think this has priority right now. If you already have a ``Gene`` class in mind, please feel free to commit something; makes it easier to discuss the interface anyway. Also, personally I would go with a simple container e.g. ``std::vector`` for now @manish7294 do you think that is reasonable?

Reply:
Sure, we can go with vector. And can always at later point of time, see what can be improved.